packages/gatsby/src/utils/get-page-data.ts

```diff
@@ -0,0 +1,61 @@
+import * as path from "path"
+import { store, emitter } from "../redux"
+import { IClearPendingPageDataWriteAction } from "../redux/types"
+import * as actions from "../redux/actions/internal"
+import {
+  IPageDataWithQueryResult,
+  readPageData as readPageDataUtil,
+} from "./page-data"
+
+export interface IPageQueryResult {
+  id: string
+  result: IPageDataWithQueryResult
+}
+
+export async function getPageData(pagePath: string): Promise<IPageQueryResult> {
+  const { queries } = store.getState()
+
+  const query = queries.trackedQueries.get(pagePath)
+
+  if (!query) {
+    throw new Error(`Could not find query ${pagePath}`)
+  }
+  if (query.running !== 0) {
+    return waitNextPageData(pagePath)
+  }
+  if (query.dirty !== 0) {
+    store.dispatch(actions.queryRunRequested({ path: pagePath, isPage: true }))
+    return waitNextPageData(pagePath)
+  }
+  // Results are up-to-date
+  return readPageData(pagePath)
```

Comment:
There is one more timing thing potentially to cover here - time between finishing query running and flushing fresh `page-data` file - with current code it's possible to fall into this return while `page-data` flush is pending for given page

Reply:
This is the current approach we are going to take for now https://github.com/gatsbyjs/gatsby/pull/28028/commits/50a9653c9ae07422268eb469353b53c207da5003

The downside of it is that page data write may be pending not because data is stale but because upstream static query hashes changed. In the latter case, we will be unnecessarily waiting.